Spacing Support: Add margin block support with configurable sides

aaronrobertshaw opened the pull request:
## Description

This adds margin block support with the configurable sides feature provided in https://github.com/WordPress/gutenberg/pull/30606.

The intention is to merge this into https://github.com/WordPress/gutenberg/pull/30607 so that both the updated padding support and this new margin support are consistent.

## How has this been tested?
Manually.

#### Test Instructions

Easiest means of testing these changes is to checkout https://github.com/WordPress/gutenberg/pull/30609 and follow its test instructions.

## Screenshots <!-- if applicable -->

<img src="https://user-images.githubusercontent.com/60436221/114001030-e8a03400-989e-11eb-8b94-00aa07f35a68.gif" width="400"/>

## Types of changes
New feature. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Add margin block support with configurable sides

aaronrobertshaw committed:
Update spacing support docs for margin

aaronrobertshaw commented:
This has been rebased after the [changes on trunk to padding.js and spacing-panel.js](https://github.com/WordPress/gutenberg/commit/f818f1d7562e87ca551777bb4c286e29b1d8f617) fixing custom CSS units.

oandregal commented:
I see this PR has all checks green (including the approval) but I don't see anyone's approval on the reviewers' roster. Did we break GitHub?

Going to give it a shot anyway. Feel free to merge if there was a previous approval. If I find anything worth fixing, we can always prepare follow-ups.

youknowriad commented:
@nosolosw I think it's because it's not targeting trunk yet.

oandregal commented on packages/edit-site/src/components/sidebar/spacing-panel.js:
```diff
@@ -123,6 +130,15 @@ export default function SpacingPanel( { context, getStyle, setStyle } ) {
 	};
 
+	const marginValues = getStyle( name, 'margin' );
+	const themeMarginValues = useThemeValues( name, 'margin' );
```
I understand this has not been introduced in this PR, so happy to iterate on a different one.

I understand that the goal of `useThemeValues( name, property )` is to get the base values (core+theme). I believe is in our best interest to evolve getStyles to also allow retrieving these data such as: `getStyle( name, property, 'theme' )` instead of doing what we do at the moment. The rationale for that is that we want to centralize access to the shape of the data, so consumers don't need to deal with shape changes, migrations, etc. ― instead, there's a centralized place for all of that (global styles provider, getStyle method, etc).

youknowriad replied:
Yes, commented about it here https://github.com/WordPress/gutenberg/pull/30607#discussion_r617366595

oandregal replied:
Oh, I saw your other comment now. So it looks like I should have started by reviewing that one.

aaronrobertshaw replied:
Thanks 👍 

`useThemeValues` has been removed and values are now simply unset. I'll create a new PR soon to update `getStyle` to support a `theme` origin.

oandregal commented:
I left a comment at https://github.com/WordPress/gutenberg/pull/30608#discussion_r618268847 Other than that, it looks fine. Though that concern doesn't seem to belong in this PR, as I've just realized this wasn't based on `trunk` but on https://github.com/WordPress/gutenberg/pull/3060 so I should have started with that one.

youknowriad commented:
Should we add the `!important` styles from my original margin block support PR? I think it makes sense.

aaronrobertshaw committed:
Add important to auto layout margins

aaronrobertshaw commented:
> Should we add the !important styles from my original margin block support PR? I think it makes sense.

I've added the `!important` to the layout styles as per your original PR and rebased the PR.

youknowriad reviewed (approved):
This is looking good to me.

I'd appreciate some extra testing if possible, maybe @jasmussen @MaggieCabrera but I guess we could enable it in Group block to ease testing here.

jasmussen commented:
Margins and paddings are so needed, I'm happy to see the feature land. 

My only concern is that we are still not really taking a stab at the more generalized efforts in https://github.com/WordPress/gutenberg/issues/27331, which see both margins and paddings grouped in a dedicated "Dimensions" panels, such as this:

<img width="700" alt="Screenshot 2021-05-07 at 10 07 14" src="https://user-images.githubusercontent.com/1204802/117419030-68600380-af1c-11eb-9c9e-12760d9485c0.png">

One aspect there as well, is that margin and padding is a very obtuse control for folks that aren't web-designers, and we might even want to use a segmented control to group the two. 

Probably margin is such an important tool that we shouldn't hold it up. But should we at least start to move towards a single "Dimensions" panel?

MaggieCabrera commented:
This is a very welcome PR, thank you for your work!! I tested this out (using #30609) and it worked as intended. +1 on getting this on the group block, it's where we most need it.
I was not able to change the default values for the separator block using theme.json though.

aaronrobertshaw commented:
> My only concern is that we are still not really taking a stab at the more generalized efforts in #27331, which see both margins and paddings grouped in a dedicated "Dimensions" panels, such as this:

I'd be happy to help make steps towards this happening though unfortunately I'm not very clear on any progress, or the plan to integrate G2 components into Gutenberg as per the screenshot. Similar issues arise with the border support panel refinements requested in https://github.com/WordPress/gutenberg/issues/31337.

> Probably margin is such an important tool that we shouldn't hold it up. 

There are a number of issues blocked by not having the ability to configure margins. Getting this one merged and then iterating on the UI would definitely get my vote.

jasmussen commented:
> or the plan to integrate G2 components into Gutenberg as per the screenshot

The designs in #27331 are not reliant on any G2 components being added to the block editor. The system is entirely compoment agnostic, and is purely an expression of which design tools are surfaced in which panels, according to block supports.

aaronrobertshaw commented:
> The designs in #27331 are not reliant on any G2 components being added to the block editor. The system is entirely compoment agnostic, and is purely an expression of which design tools are surfaced in which panels, according to block supports.

Thanks for the additional info. I'll look into it. Perhaps I was taking the design screenshots too literally.

jasmussen commented:
> Perhaps I was taking the design screenshots too literally.

I mean, we'll need some padding, margin, CSS, styling updates, and ideally we have a system that helps float/flex/rearrange/correctly place the form controls based on what's supported and so on, in a 2 or 3 column system. But there are steps that can be taken before that.

aaronrobertshaw commented:
> I was not able to change the default values for the separator block using theme.json though.

Thanks @MaggieCabrera, you should be able to test this margin support via the theme.json for the separator block now. 

In case it helps, here is the [theme.json](https://gist.github.com/aaronrobertshaw/f528797fc55d95b947315e1862d4a204) I used while testing.

MaggieCabrera commented:
> Thanks @MaggieCabrera, you should be able to test this margin support via the theme.json for the separator block now.
> 
> In case it helps, here is the [theme.json](https://gist.github.com/aaronrobertshaw/f528797fc55d95b947315e1862d4a204) I used while testing.

Yeah, I was able to see the spacing menu on the editor, what I meant was that I couldn't do something like this on my theme.json file:

```
			"core/separator": {
				"spacing": {
					"margin": {
						"top": "200px"
					}
				}
			},
```

This is going to be important for certain blocks so that the theme can set some good default values and we can actually do it for the padding right now.

aaronrobertshaw commented:
@MaggieCabrera sorry if I wasn't clear in my comment. I meant you could now do exactly what you were expecting in setting the margins within the theme.json styles. The linked gist actually set the top and bottom margins.

```json
	"styles": {
		"blocks": {
			"core/separator": {
				"spacing": {
					"margin": {
						"top": "100px",
						"bottom": "100px"
					}
				}
			}
		}
	}
```

Please let me know if there are any other issues you were facing.

MaggieCabrera commented:
> @MaggieCabrera sorry if I wasn't clear in my comment. I meant you could now do exactly what you were expecting in setting the margins within the theme.json styles. The linked gist actually set the top and bottom margins.

Yeah you are totally right, it was a theme issue not letting me see the changes. I tested again with your theme.json file using [emptytheme](https://github.com/WordPress/theme-experiments/tree/master/emptytheme) and I can see the margins on the frontend but the editor shows 15px margin on the separators isntead of 100px: 

Editor | Frontend
--- | ---
<img width="779" alt="Screenshot 2021-05-10 at 09 56 13" src="https://user-images.githubusercontent.com/3593343/117625262-229e7780-b176-11eb-85c9-cad5cb993e55.png"> | <img width="1060" alt="Screenshot 2021-05-10 at 09 56 02" src="https://user-images.githubusercontent.com/3593343/117625264-23370e00-b176-11eb-9886-d7fd6a3f9352.png">

aaronrobertshaw commented:
>  I can see the margins on the frontend but the editor shows 15px margin on the separators instead of 100px:

There was a 15px default value applied to the separator block specifically that was overriding the theme value. That was removed in https://github.com/WordPress/gutenberg/pull/30609/commits/0f575539004a2e8c667ed3c8fb38853399c74d9a before I advised it should be right to test again. 

Can you confirm if you pulled the latest changes from https://github.com/WordPress/gutenberg/pull/30609 before re-testing?

This is what I see when loading the editor with the theme.json gist linked above, no defaults in the control.

<img width="600" alt="Screen Shot 2021-05-10 at 6 19 59 pm" src="https://user-images.githubusercontent.com/60436221/117628397-72079680-b1bc-11eb-8286-e5721f606fc8.png">

aaronrobertshaw commented:
It's also worth mentioning there isn't a means at present to determine the default merged ( default, theme & user ) style value to populate the block support controls within the block editor, only the full site editor. 

If we wish to address that, it's an issue for all current block support controls and not unique to this PR. A separate PR would be best to discuss that further if desired.

MaggieCabrera commented:
> Can you confirm if you pulled the latest changes from #30609 before re-testing?
> 

Yes, I pulled the latest from the branch and it works nicely, awesome!! Thanks for your help and for the PR, can't wait to be able to play this :D

youknowriad commented:
I think the dimensions UI redesign can be considered a follow-up.

aaronrobertshaw commented:
I've rebased again to resolve a conflict. Should be ok to merge once tests pass. Then we can iterate on the dimensions UI redesign as suggested.
